lib/samson/readonly_db.rb

```diff
@@ -0,0 +1,48 @@
+# frozen_string_literal: true
+
+# Make Activerecord readonly by blocking write requests
+# NOTE: not perfect and can be circumvented with multiline sql statements or `;`
+module Samson
+  module ReadonlyDb
+    ALLOWED = ["SELECT ", "SHOW ", "SET  @@SESSION.", "EXPLAIN "].freeze
+    PROMPT_CHANGE = ["(", "(readonly "].freeze
+    PROMPTS = [:PROMPT_I, :PROMPT_N].freeze
+
+    class << self
+      def enable
+        return if @subscriber
+        @subscriber = ActiveSupport::Notifications.subscribe("sql.active_record") do |*, payload|
```

Comment:
is there any documented guarantees that the notification will fire before the actual sql gets executed?